packages/preferences/src/browser/abstract-resource-preference-provider.ts

```diff
@@ -80,7 +83,7 @@ export abstract class AbstractResourcePreferenceProvider extends PreferenceProvi
     protected abstract getScope(): PreferenceScope;
 
     protected get valid(): boolean {
-        return this.model && this.model.valid || false;
+        return this.modelInitialized ? !!this.model?.valid : !!Object.keys(this.preferences).length;
```

Comment:
if we're checking `modelInitialized`, can't we write `model!.valid`? Also, wouldn't `Object.keys(this.preferences).length !== 0` be clearer? Because that's what it checks, no?

Reply:
Also, could you explain the scenario that makes `modelInitialized` necessary? I suspect it has to do with the `loading' promise, but I don't see it.

Reply:
In the new code, there are three states we need to account for:
 - Before we've initialized the model -> We only care about whether we have anything to report, so check `this.preferences`.
 - Once we've initialized the model and before we dispose -> `this.model` should be defined and valid
 - After disposed -> `this.model` will be `undefined`, so we can't write `model!.valid`, but it's still the relevant check, so we need `this.modelInitialized` to push us to check `this.model` rather than having something like `this.model?.valid || Object.keys(this.preferences) > 0`.

Re: the `loading` promise vs. `this.modelInitialized`, various methods that are currently synchronous refer to `this.valid`, so we need a synchronous way to decide whether or not to check the `model`, rather than using the promise and changing to async everywhere. In my implementation with full lazy initiation, the loading promise goes away completely

Reply:
> 
> @tsmaeder, any thoughts?

Sorry, that whizzed by me. Yeah, no objections. Incremental improvements for teh win!